File: python/src/circle_ai/skills/file_skill_store.py
# file_skill_store.py
#
# Port of CircleAI.Skills FileSkillStore.cs (C# — the EXACT spec).
#
# ISkillStore backed by SKILL.md files in a directory. Each file uses YAML
# front-matter for metadata and a Markdown body for the skill instructions —
# the same format used by Hermes OS1. The ``id`` front-matter field is optional;
# when absent the file name (without extension) is used as the skill ID.
#
# Expected file format:
#
#   ---
#   id: calendar-summariser
#   name: Calendar Summariser
#   description: Summarises upcoming calendar events into a concise digest
#   tags: [productivity, calendar, summaries]
#   ---
#
#   ## Instructions
#   When the user asks about their schedule, call the calendar tool...

from __future__ import annotations

import os
from datetime import datetime, timezone
from typing import Dict, List, Optional

from .contracts import ISkillStore, SkillDetail, SkillDraft, SkillSource, SkillSummary
from .in_memory_skill_store import InMemorySkillStore


def _require_not_whitespace(value: str, name: str) -> None:
    if value is None or value.strip() == "":
        raise ValueError(f"{name} must not be null or whitespace")


class FileSkillStore(ISkillStore):
    """:class:`ISkillStore` backed by SKILL.md files in a directory."""

    def __init__(self, directory_path: str) -> None:
        _require_not_whitespace(directory_path, "directory_path")
        self._directory_path = directory_path
        os.makedirs(self._directory_path, exist_ok=True)

    async def list_async(self, cancellation_token: Optional[object] = None) -> List[SkillSummary]:
        results: List[SkillSummary] = []
        for file in self._get_skill_files():
            detail = self._read_skill_file(file)
            if detail is not None:
                results.append(self._to_summary(detail))
        results.sort(key=lambda s: s.name.casefold())
        return results

    async def get_async(self, id: str, cancellation_token: Optional[object] = None) -> Optional[SkillDetail]:
        _require_not_whitespace(id, "id")
        for file in self._get_skill_files():
            detail = self._read_skill_file(file)
            if detail is not None and detail.id.casefold() == id.casefold():
                return detail
        return None

    async def search_async(self, query: str, cancellation_token: Optional[object] = None) -> List[SkillSummary]:
        if query is None or query.strip() == "":
            return []
        q = query.strip()
        results: List[SkillSummary] = []
        for file in self._get_skill_files():
            detail = self._read_skill_file(file)
            if detail is not None and self._matches_query(detail, q):
                results.append(self._to_summary(detail))
        results.sort(key=lambda s: s.name.casefold())
        return results

    async def upsert_async(
        self, id: Optional[str], draft: SkillDraft, cancellation_token: Optional[object] = None
    ) -> SkillDetail:
        if draft is None:
            raise ValueError("draft must not be None")

        effective_id = (
            InMemorySkillStore.generate_slug(draft.name) if (id is None or id.strip() == "") else id.strip()
        )

        file_path = os.path.join(self._directory_path, f"{effective_id}.md")
        if draft.tags is not None and len(draft.tags) > 0:
            tags = "[" + ", ".join(draft.tags) + "]"
        else:
            tags = "[]"

        lines = [
            "---",
            f"id: {effective_id}",
            f"name: {draft.name}",
            f"description: {draft.description}",
            f"tags: {tags}",
            "---",
            "",
        ]
        # C# AppendLine emits a trailing newline after each header line and the
        # blank separator, then Append (no newline) writes the body.
        content = "\n".join(lines) + "\n" + draft.instructions

        with open(file_path, "w", encoding="utf-8") as fh:
            fh.write(content)

        return SkillDetail(
            id=effective_id,
            name=draft.name,
            description=draft.description,
            instructions=draft.instructions,
            tags=list(draft.tags) if draft.tags is not None else [],
            source=SkillSource.File,
            last_modified=datetime.now(timezone.utc),
        )

    async def delete_async(self, id: str, cancellation_token: Optional[object] = None) -> None:
        _require_not_whitespace(id, "id")
        file_path = os.path.join(self._directory_path, f"{id}.md")
        if os.path.exists(file_path):
            os.remove(file_path)

    # ── Parsing ────────────────────────────────────────────────────────────────

    def _get_skill_files(self) -> List[str]:
        try:
            names = os.listdir(self._directory_path)
        except OSError:
            return []
        return [
            os.path.join(self._directory_path, n)
            for n in names
            if n.lower().endswith(".md") and os.path.isfile(os.path.join(self._directory_path, n))
        ]

    @staticmethod
    def _read_skill_file(file_path: str) -> Optional[SkillDetail]:
        try:
            with open(file_path, "r", encoding="utf-8") as fh:
                content = fh.read()
        except OSError:
            return None
        file_name_no_ext = os.path.splitext(os.path.basename(file_path))[0]
        return FileSkillStore.parse_skill_file(content, file_name_no_ext, file_path)

    @staticmethod
    def parse_skill_file(content: str, file_name_without_ext: str, file_path: str) -> Optional[SkillDetail]:
        if content is None or content.strip() == "":
            return None

        # Locate the YAML front-matter block between the first two "---" lines.
        lines = content.replace("\r\n", "\n").split("\n")
        if len(lines) < 2 or lines[0].strip() != "---":
            return None

        front_matter_end = -1
        for i in range(1, len(lines)):
            if lines[i].strip() == "---":
                front_matter_end = i
                break
        if front_matter_end < 0:
            return None

        # Parse front-matter key: value pairs (case-insensitive keys).
        meta: Dict[str, str] = {}
        for i in range(1, front_matter_end):
            line = lines[i]
            colon = line.find(":")
            if colon < 0:
                continue
            key = line[:colon].strip().casefold()
            value = line[colon + 1 :].strip()
            meta[key] = value

        id_val = meta.get("id")
        skill_id = id_val if (id_val is not None and id_val.strip() != "") else file_name_without_ext
        name = meta.get("name", skill_id)
        description = meta.get("description", "")
        tags = FileSkillStore._parse_tags_list(meta.get("tags", ""))

        # Everything after the closing "---" is the instructions body.
        instructions = "\n".join(lines[front_matter_end + 1 :]).strip()

        if os.path.exists(file_path):
            last_modified = datetime.fromtimestamp(os.path.getmtime(file_path), tz=timezone.utc)
        else:
            last_modified = datetime.now(timezone.utc)

        return SkillDetail(skill_id, name, description, instructions, tags, SkillSource.File, last_modified)

    @staticmethod
    def _parse_tags_list(raw: str) -> List[str]:
        """Parse a YAML inline list like ``[a, b, c]`` or a bare scalar."""
        if raw is None or raw.strip() == "":
            return []
        raw = raw.strip()
        if raw.startswith("[") and raw.endswith("]"):
            raw = raw[1:-1]
        return [t.strip() for t in raw.split(",") if t.strip() != ""]

    @staticmethod
    def _to_summary(d: SkillDetail) -> SkillSummary:
        return SkillSummary(d.id, d.name, d.description, d.tags, d.source)

    @staticmethod
    def _matches_query(s: SkillDetail, query: str) -> bool:
        q = query.casefold()
        return (
            q in s.name.casefold()
            or q in s.description.casefold()
            or any(q in t.casefold() for t in s.tags)
        )

File: src/CircleAI.Skills/FileSkillStore.cs
using System.Text;

namespace CircleAI.Skills;

/// <summary>
/// <see cref="ISkillStore"/> backed by SKILL.md files in a directory.
/// Each file uses YAML front-matter for metadata and Markdown body for
/// the skill instructions — the same format used by Hermes OS1.
/// </summary>
/// <remarks>
/// <para>Expected file format:</para>
/// <code>
/// ---
/// id: calendar-summariser
/// name: Calendar Summariser
/// description: Summarises upcoming calendar events into a concise digest
/// tags: [productivity, calendar, summaries]
/// ---
///
/// ## Instructions
/// When the user asks about their schedule, call the calendar tool…
/// </code>
/// <para>
/// The <c>id</c> front-matter field is optional; when absent the file name
/// (without extension) is used as the skill ID.
/// </para>
/// </remarks>
public sealed class FileSkillStore : ISkillStore
{
    private readonly string _directoryPath;

    /// <summary>
    /// Initialises the store. Creates <paramref name="directoryPath"/> if it
    /// does not yet exist.
    /// </summary>
    public FileSkillStore(string directoryPath)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(directoryPath);
        _directoryPath = directoryPath;
        Directory.CreateDirectory(_directoryPath);
    }

    /// <inheritdoc />
    public async Task<IReadOnlyList<SkillSummary>> ListAsync(CancellationToken cancellationToken = default)
    {
        var results = new List<SkillSummary>();
        foreach (var file in GetSkillFiles())
        {
            cancellationToken.ThrowIfCancellationRequested();
            var detail = await ReadSkillFileAsync(file, cancellationToken).ConfigureAwait(false);
            if (detail is not null)
                results.Add(ToSummary(detail));
        }
        return results.OrderBy(s => s.Name, StringComparer.OrdinalIgnoreCase).ToList();
    }

    /// <inheritdoc />
    public async Task<SkillDetail?> GetAsync(string id, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        foreach (var file in GetSkillFiles())
        {
            cancellationToken.ThrowIfCancellationRequested();
            var detail = await ReadSkillFileAsync(file, cancellationToken).ConfigureAwait(false);
            if (detail is not null && string.Equals(detail.Id, id, StringComparison.OrdinalIgnoreCase))
                return detail;
        }
        return null;
    }

    /// <inheritdoc />
    public async Task<IReadOnlyList<SkillSummary>> SearchAsync(string query, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(query))
            return Array.Empty<SkillSummary>();

        var q = query.Trim();
        var results = new List<SkillSummary>();
        foreach (var file in GetSkillFiles())
        {
            cancellationToken.ThrowIfCancellationRequested();
            var detail = await ReadSkillFileAsync(file, cancellationToken).ConfigureAwait(false);
            if (detail is not null && MatchesQuery(detail, q))
                results.Add(ToSummary(detail));
        }
        return results.OrderBy(s => s.Name, StringComparer.OrdinalIgnoreCase).ToList();
    }

    /// <inheritdoc />
    public async Task<SkillDetail> UpsertAsync(string? id, SkillDraft draft, CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(draft);

        var effectiveId = string.IsNullOrWhiteSpace(id)
            ? InMemorySkillStore.GenerateSlug(draft.Name)
            : id.Trim();

        var filePath = Path.Combine(_directoryPath, $"{effectiveId}.md");
        var tags = draft.Tags is { Count: > 0 }
            ? $"[{string.Join(", ", draft.Tags)}]"
            : "[]";

        var content = new StringBuilder();
        content.AppendLine("---");
        content.AppendLine($"id: {effectiveId}");
        content.AppendLine($"name: {draft.Name}");
        content.AppendLine($"description: {draft.Description}");
        content.AppendLine($"tags: {tags}");
        content.AppendLine("---");
        content.AppendLine();
        content.Append(draft.Instructions);

        await File.WriteAllTextAsync(filePath, content.ToString(), Encoding.UTF8, cancellationToken)
            .ConfigureAwait(false);

        return new SkillDetail(
            Id: effectiveId,
            Name: draft.Name,
            Description: draft.Description,
            Instructions: draft.Instructions,
            Tags: draft.Tags ?? Array.Empty<string>(),
            Source: SkillSource.File,
            LastModified: DateTimeOffset.UtcNow);
    }

    /// <inheritdoc />
    public Task DeleteAsync(string id, CancellationToken cancellationToken = default)
    {
        ArgumentException.ThrowIfNullOrWhiteSpace(id);
        var filePath = Path.Combine(_directoryPath, $"{id}.md");
        if (File.Exists(filePath))
            File.Delete(filePath);
        return Task.CompletedTask;
    }

    // ------------------------------------------------------------------
    // Parsing
    // ------------------------------------------------------------------

    private IEnumerable<string> GetSkillFiles() =>
        Directory.EnumerateFiles(_directoryPath, "*.md", SearchOption.TopDirectoryOnly);

    private static async Task<SkillDetail?> ReadSkillFileAsync(
        string filePath, CancellationToken ct)
    {
        string content;
        try
        {
            content = await File.ReadAllTextAsync(filePath, Encoding.UTF8, ct).ConfigureAwait(false);
        }
        catch
        {
            return null;
        }

        return ParseSkillFile(content, Path.GetFileNameWithoutExtension(filePath), filePath);
    }

    public static SkillDetail? ParseSkillFile(string content, string fileNameWithoutExt, string filePath)
    {
        if (string.IsNullOrWhiteSpace(content)) return null;

        // Locate the YAML front-matter block between the first two "---" lines.
        var lines = content.Replace("\r\n", "\n").Split('\n');
        if (lines.Length < 2 || lines[0].Trim() != "---")
            return null;

        var frontMatterEnd = -1;
        for (int i = 1; i < lines.Length; i++)
        {
            if (lines[i].Trim() == "---") { frontMatterEnd = i; break; }
        }
        if (frontMatterEnd < 0) return null;

        // Parse front-matter key: value pairs.
        var meta = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        for (int i = 1; i < frontMatterEnd; i++)
        {
            var line = lines[i];
            var colon = line.IndexOf(':');
            if (colon < 0) continue;
            var key = line[..colon].Trim();
            var value = line[(colon + 1)..].Trim();
            meta[key] = value;
        }

        var id = meta.TryGetValue("id", out var idVal) && !string.IsNullOrWhiteSpace(idVal)
            ? idVal
            : fileNameWithoutExt;
        var name = meta.TryGetValue("name", out var nameVal) ? nameVal : id;
        var description = meta.TryGetValue("description", out var descVal) ? descVal : string.Empty;
        var tags = ParseTagsList(meta.TryGetValue("tags", out var tagsVal) ? tagsVal : string.Empty);

        // Everything after the closing "---" is the instructions body.
        var instructionsLines = lines.Skip(frontMatterEnd + 1);
        var instructions = string.Join("\n", instructionsLines).Trim();

        var lastModified = File.Exists(filePath)
            ? new DateTimeOffset(File.GetLastWriteTimeUtc(filePath), TimeSpan.Zero)
            : DateTimeOffset.UtcNow;

        return new SkillDetail(id, name, description, instructions, tags, SkillSource.File, lastModified);
    }

    /// <summary>
    /// Parses a YAML inline list like <c>[a, b, c]</c> or a bare scalar.
    /// </summary>
    private static IReadOnlyList<string> ParseTagsList(string raw)
    {
        if (string.IsNullOrWhiteSpace(raw)) return Array.Empty<string>();
        raw = raw.Trim();
        if (raw.StartsWith('[') && raw.EndsWith(']'))
            raw = raw[1..^1];
        return raw.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
                  .Where(t => !string.IsNullOrWhiteSpace(t))
                  .ToArray();
    }

    private static SkillSummary ToSummary(SkillDetail d) =>
        new(d.Id, d.Name, d.Description, d.Tags, d.Source);

    /// <summary>Does this skill answer <paramref name="query"/>?</summary>
    /// <remarks>
    /// ID first — it is the handle the compact listing hands out, so it has to be
    /// the one thing a lookup is guaranteed to find. Kept identical to the other
    /// stores: a skill that resolves in memory and not on disk is a bug that only
    /// shows up on someone's phone.
    /// </remarks>
    private static bool MatchesQuery(SkillDetail s, string query) =>
        s.Id.Contains(query, StringComparison.OrdinalIgnoreCase) ||
        s.Name.Contains(query, StringComparison.OrdinalIgnoreCase) ||
        s.Description.Contains(query, StringComparison.OrdinalIgnoreCase) ||
        s.Tags.Any(t => t.Contains(query, StringComparison.OrdinalIgnoreCase));
}
